### PR TITLE
feat[shadows]: Added Shadow Mapping

### PR DESCRIPTION
**Pull request by MacXxs**
**To toggle shadows use F3**

* Implemented Shadow Mapping for global shadows
* Implemented shadow bias to fix peter panning and shadow acne
* Implemented SDSM to improve shadow resolution using depth map parallel reduction
* Added Compute shaders to the engine in order to do the parallel reduction and increase performance using gpu parallelization
* Added new functions to obtain objects in a scene given a frustum and implemented new functions for Batching in order to use shadows
* Started the refactor for batching

**TODO: We will need to implement a way of keeping the stencil buffer from the opaque drawing in order to fix the highlight**

### CODE REVIEW

**Review by AlexisDieu — APPROVED** [review]

File: Source/DataModels/Scene/Scene.cpp [hunk 1]
```diff
@@ -114,6 +114,159 @@ bool Scene::IsInsideACamera(const AABB& aabb) const
...
+
+#ifdef ENGINE
+	CalculateNonStaticObjectsInFrustum(frustum, App->GetModule<ModuleScene>()->GetSelectedGameObject(), 
+									   objectsInFrustum);
+#endif
+
+	return objectsInFrustum;
+}
+
+void Scene::CalculateObjectsInFrustum(const math::Frustum* frustum, const Quadtree* quad,
```
Comment (PaablooCH) [inline Source/DataModels/Scene/Scene.cpp:136]: does this function the same as the one in moduleRender?

File: Source/Modules/ModuleRender.cpp [hunk 2]
```diff
@@ -483,7 +539,27 @@ void ModuleRender::WindowResized(unsigned width, unsigned height)
 
 void ModuleRender::UpdateBuffers(unsigned width, unsigned height)
 {
-	gBuffer->InitGBuffer(width, height);
+	gBuffer->InitGBuffer(width,height);
```
Comment (PaablooCH) [inline Source/Modules/ModuleRender.cpp:542]: gBuffer->InitGBuffer(width, height);

File: Source/Modules/ModuleRender.cpp [hunk 3]
```diff
@@ -483,7 +539,27 @@ void ModuleRender::WindowResized(unsigned width, unsigned height)
...
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
+
+	glFramebufferTexture2D(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_TEXTURE_2D, gShadowMap, 0);
+
+	glDrawBuffer(GL_NONE);
+	glReadBuffer(GL_NONE);
+
+	if (glCheckFramebufferStatus(GL_FRAMEBUFFER) != GL_FRAMEBUFFER_COMPLETE)
+	{
+		LOG_ERROR("G Framebuffer not complete!");
```
Comment (PaablooCH) [inline Source/Modules/ModuleRender.cpp:561]: Change the error message

File: Source/Modules/ModuleRender.cpp [hunk 4]
```diff
@@ -547,6 +623,13 @@ void ModuleRender::UpdateBuffers(unsigned width, unsigned height)
 	}
 		
 	glBindFramebuffer(GL_FRAMEBUFFER, 0);
+
+	glBindTexture(GL_TEXTURE_2D, parallelReductionInTexture);
```
Comment (PaablooCH) [inline Source/Modules/ModuleRender.cpp:627]: where are you binding this texture?

**Review by PaablooCH — COMMENTED** [review]
to finish:

File: Source/DataModels/Batch/GeometryBatch.cpp [hunk 5]
```diff
@@ -838,6 +838,143 @@ void GeometryBatch::BindBatch(bool selected)
...
+		FillMaterial();
+	}
+
+	std::vector<Command> commands;
+	commands.reserve(componentsInBatch.size());
+
+	int drawCount = 0;
+
+	GameObject* selectedGo = App->GetModule<ModuleScene>()->GetSelectedGameObject();
+	bool isRoot = selectedGo->GetParent() == nullptr;
```
[the remark this answers is not in the document]
Comment (MacXxs) [reply Source/DataModels/Batch/GeometryBatch.cpp:901]: Gonna remove it in the AnimatedTextures PR

File: Source/DataModels/Batch/GeometryBatch.cpp [hunk 6]
```diff
@@ -838,6 +838,143 @@ void GeometryBatch::BindBatch(bool selected)
...
+
+	for (ComponentMeshRenderer* component : componentsInBatch)
+	{
+		assert(component);
+
+		std::vector<GameObject*>::iterator it = std::find(objects.begin(), objects.end(), component->GetOwner());
+		
+		bool draw = it != objects.end();
+		
+		if (it != objects.end())
```
[the remark this answers is not in the document]
Comment (MacXxs) [reply Source/DataModels/Batch/GeometryBatch.cpp:911]: Gonna push it in the AnimatedTextures PR

File: Source/DataModels/Components/ComponentDirLight.h [hunk 7]
```diff
@@ -15,6 +15,10 @@ class ComponentDirLight : public ComponentLight
 
 	void Draw() const override;
 
+	void OnTransformChanged() override;
+
+	float2 shadowBias = float2(0.006, 0.0009);
```
[the remark this answers is not in the document]
Comment (MacXxs) [reply Source/DataModels/Components/ComponentDirLight.h:20]: Gonna push it in the AnimatedTextures PR

File: Source/DataModels/Scene/Scene.h [hunk 8]
```diff
@@ -42,6 +42,14 @@ class Scene
 	bool IsInsideACamera(const OBB& obb) const;
 	bool IsInsideACamera(const AABB& aabb) const;
 
+	std::vector<GameObject*> ObtainObjectsInFrustum(const math::Frustum* frustum);
+	void CalculateObjectsInFrustum(const math::Frustum* frustum, const Quadtree* quad, 
+								   std::vector<GameObject*>& gos);
+	void CalculateNonStaticObjectsInFrustum(const math::Frustum* frustum, GameObject* go,
+										    std::vector<GameObject*>& gos);
+	bool frustumInQuadTree(const math::Frustum* frustum, const Quadtree* quad);
```
[the remark this answers is not in the document]
Comment (MacXxs) [reply Source/DataModels/Scene/Scene.h:50]: Gonna push it in the AnimatedTextures PR

File: Source/DataModels/Scene/Scene.h [hunk 9]
```diff
@@ -42,6 +42,14 @@ class Scene
...
+	bool objectInFrustum(const math::Frustum* frustum, const AABB& aabb);
```
[the remark this answers is not in the document]
Comment (MacXxs) [reply Source/DataModels/Scene/Scene.h:51]: Gonna push it in the AnimatedTextures PR

File: Source/DataModels/Scene/Scene.cpp (hunk 1 above)
Comment (MacXxs) [reply Source/DataModels/Scene/Scene.cpp:136]: No, this one is a generic function that given any frustum, it returns all objects (both static and non static) inside said frustum.
The one in moduleRender only uses the engine camera and returns only static objects. Furthermore the FillRenderList function will be removed with the final PR for the refactor of the batching and instead, this one will be used.

File: Source/Modules/ModuleRender.cpp (hunk 4 above)
Comment (MacXxs) [reply Source/Modules/ModuleRender.cpp:627]: They are not attached to a frame buffer, and are being binded in order to modify its storage as you can see by the following lines.

File: Source/Modules/ModuleRender.cpp (hunk 3 above)
Comment (MacXxs) [reply Source/Modules/ModuleRender.cpp:561]: Gonna push it in the AnimatedTextures PR

File: Source/Modules/ModuleRender.cpp (hunk 2 above)
Comment (MacXxs) [reply Source/Modules/ModuleRender.cpp:542]: Gonna push it in the AnimatedTextures PR

File: Source/Modules/ModuleRender.cpp [hunk 10]
```diff
@@ -135,6 +140,14 @@ ModuleRender::~ModuleRender()
 	delete gBuffer;
 	objectsInFrustrumDistances.clear();
 	gameObjectsInFrustrum.clear();
+
+	glDeleteFramebuffers(2, &frameBuffer[0]);
+	glDeleteTextures(2, &renderedTexture[0]);
+
+	glDeleteRenderbuffers(1, &depthStencilRenderBuffer);
```
[the remark this answers is not in the document]
Comment (MacXxs) [reply Source/Modules/ModuleRender.cpp:147]: Gonna push it in the AnimatedTextures PR

File: Source/Modules/ModuleRender.cpp [hunk 11]
```diff
@@ -20,6 +23,8 @@
 #include "DataModels/Skybox/Skybox.h"
 #include "DataModels/GBuffer/GBuffer.h"
 
+#include "debugdraw.h"
```
[the remark this answers is not in the document]
Comment (MacXxs) [reply Source/Modules/ModuleRender.cpp:26]: Gonna push it in the AnimatedTextures PR

**Pull request merged**